wandb/sdk/lib/console_capture.py

```diff
@@ -0,0 +1,169 @@
+"""Module for intercepting stdout/stderr.
+
+This patches the `write()` method of `stdout` and `stderr` on import.
+Once patched, it is not possible to unpatch or repatch, though individual
+callbacks can be removed.
+
+We assume that all other writing methods on the object delegate to `write()`,
+like `writelines()`. This is not guaranteed to be true, but it is true for
+common implementations. In particular, CPython's implementation of IOBase's
+`writelines()` delegates to `write()`.
+
+It is important to note that this technique interacts poorly with other
+code that performs similar patching if it also allows unpatching as this
+discards our modification. This is why we patch on import and do not support
+unpatching:
+
+    with contextlib.redirect_stderr(...):
+        from ... import console_capture
+        # Here, everything works fine.
+    # Here, callbacks are never called again.
+"""
+
+from __future__ import annotations
+
+import sys
+import threading
+from typing import IO, AnyStr, Callable, Protocol
+
+
+class CannotCaptureConsoleError(Exception):
```

Comment:
i don't remember what we decide about using the wandb base exception, my intuition is to use to allow a general catch all for any wandb related exception, but i think maybe you had a case for why not?

Reply:
I think we should only use it for intentional exceptions that the user is meant to see or catch. This exception is meant to be caught by our code.